### PR TITLE
Allow platform variants for Windows plugins

### LINKED ISSUE #82815: Windows plugins need to be able to distinguish the Windows platform(s) they support
The plan for UWP plugins is to share the same implementation folder, with the expectation that many plugins will be able to be written such that they share a significant amount of code between the two (possibly all the code, as Project Reunion evolves). Not all plugins will support both Win32 and UWP, however, so we need a way for plugins to specify which they support.

Importantly during this initial phase it's likely that none do (it's possible that some work as-is by accident), so currently trying to build a project that uses any Windows-supporting plugins in UWP mode will almost certainly fail to compile. Adding a way to specify modes (defaulting to Win32 when not provided, for backwards compatibility) will address both this, and the long-term need.

### PR DESCRIPTION
**Pull request by stuartmorgan-g**
Windows plugins are designed to share implementations between Win32 and
UWP, but not all plugins will support both. This adds a new
'supportedVariants' key to Windows plugins that allows specifying
'win32' and/or 'uwp' (and potentially others in the future in case that
becomes necessary).

Plugins without any supported variants will be assumed to be Win32 for
backward compatibility.

This will allow compiling Windows projects that use Win32-only Windows
plugins (which is currently all of them) in UWP mode. The plugins will
of course throw missing implementation exceptions at runtime, but tehy
won't prevent being able to build as they currently do.

Fixes https://github.com/flutter/flutter/issues/82815

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by stuartmorgan** [comment]
Feedback on the new schema elements welcome. I didn't do a design doc for it since it's pretty straightforward, but we can discuss here if anyone thinks it should be changed (different name, completely different structure such as a bunch of optional booleans, etc.)

**Comment by jonahwilliams** [comment]
> 'supportedVariants' key to Windows plugins that allows specifying
'win32' or 'uwp' 

Should this be `and`?

File: packages/flutter_tools/lib/src/flutter_plugins.dart [hunk 1]
```diff
@@ -101,23 +101,24 @@ const String _kFlutterPluginsNameKey = 'name';
 const String _kFlutterPluginsPathKey = 'path';
 const String _kFlutterPluginsDependenciesKey = 'dependencies';
 
-  /// Filters [plugins] to those supported by [platformKey].
-  List<Map<String, dynamic>> _filterPluginsByPlatform(List<Plugin>plugins, String platformKey) {
-    final Iterable<Plugin> platformPlugins = plugins.where((Plugin p) {
-      return p.platforms.containsKey(platformKey);
+/// Filters [plugins] to those supported by [platformKey], optionally restricted
+/// to the specific [variant].
+List<Map<String, dynamic>> _filterPluginsByPlatform(List<Plugin>plugins, String platformKey) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/flutter_plugins.dart:106]:
```suggestion
List<Map<String, dynamic>> _filterPluginsByPlatform(List<Plugin> plugins, String platformKey) {
```

File: packages/flutter_tools/lib/src/flutter_plugins.dart [hunk 2]
```diff
@@ -101,23 +101,24 @@ const String _kFlutterPluginsNameKey = 'name';
 const String _kFlutterPluginsPathKey = 'path';
 const String _kFlutterPluginsDependenciesKey = 'dependencies';
 
-  /// Filters [plugins] to those supported by [platformKey].
-  List<Map<String, dynamic>> _filterPluginsByPlatform(List<Plugin>plugins, String platformKey) {
-    final Iterable<Plugin> platformPlugins = plugins.where((Plugin p) {
-      return p.platforms.containsKey(platformKey);
+/// Filters [plugins] to those supported by [platformKey], optionally restricted
+/// to the specific [variant].
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/flutter_plugins.dart:105]: I don't see a `variant` param here

**Review by jonahwilliams — COMMENTED** [review]
Do we have an issue filled for pub support?

I'm also slightly concerned about how this scales to other platforms, if we need to worry about matching supported variants to supported platforms. I can't think of any other examples of platforms with this sort of architecture though, so maybe we could just punt on it

File: packages/flutter_tools/lib/src/flutter_plugins.dart (hunk 2 above)
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/flutter_plugins.dart:105]: Left over from a path I didn't end up using, fixed.

**Comment by stuartmorgan** [comment]
> Do we have an issue filled for pub support?

No, I figured we'd wait until we'd updated at least one plugin to make sure this does actually feel like the right approach in practice, and to provide a test case.

> I'm also slightly concerned about how this scales to other platforms, if we need to worry about matching supported variants to supported platforms. I can't think of any other examples of platforms with this sort of architecture though, so maybe we could just punt on it

Yes, I'm not sure if we'd ever use this for other things either. I did this more generically to avoid lots of type checking and casting in the higher-level logic (originally the enum was Windows-specific)—I can imagine us maybe doing alternate Linux toolkits this way, in theory? But since the code structure doesn't need to match the yaml structure I figured if we did use it more places later and this approach became confusing, we could rework the classes and enums without breaking anything.

**Comment by stuartmorgan** [comment]
> > 'supportedVariants' key to Windows plugins that allows specifying
> > 'win32' or 'uwp'
> 
> Should this be `and`?

Changed to and/or

**Comment by jonahwilliams** [comment]
Makes sense!

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by stuartmorgan** [comment]
@cbracken Did you want to weigh in on this?

File: packages/flutter_tools/lib/src/flutter_plugins.dart [hunk 3]
```diff
@@ -101,23 +101,23 @@ const String _kFlutterPluginsNameKey = 'name';
...
-    final Iterable<Plugin> platformPlugins = plugins.where((Plugin p) {
-      return p.platforms.containsKey(platformKey);
+/// Filters [plugins] to those supported by [platformKey].
+List<Map<String, dynamic>> _filterPluginsByPlatform(List<Plugin> plugins, String platformKey) {
+  final Iterable<Plugin> platformPlugins = plugins.where((Plugin p) {
+    return p.platforms.containsKey(platformKey);
+  });
+
+  final Set<String> pluginNames = platformPlugins.map((Plugin plugin) => plugin.name).toSet();
+  final List<Map<String, dynamic>> list = <Map<String, dynamic>>[];
```
Comment (cbracken) [inline packages/flutter_tools/lib/src/flutter_plugins.dart:111]: nit: maybe rename something more descriptive like `plugins` while you're in here. Not a huge improvement, but a little better.

**Review by cbracken — APPROVED** [review]
lgtm

File: packages/flutter_tools/lib/src/flutter_plugins.dart (hunk 3 above)
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/flutter_plugins.dart:111]: The bigger issue is that this method is called "filter" but what it actually does is filter and tranform. But I'll call this `pluginInfo` to at least help highlight that the list isn't actually of `Plugin`s.

Comment (cbracken) [reply packages/flutter_tools/lib/src/flutter_plugins.dart]: pluginInfo is good; originally I'd though `platformPlugins`, but that's (a) taken and (b) only really a serialised representation of which platform plugins we're talking about :/

**Pull request merged**